### PR TITLE
refactor(ndt_scan_matcher, map_loader): remove map_module

### PR DESCRIPTION
**Pull request by SakodaShintaro**
## Description

This pull request proposes removing map_module. I think this is generally a good change, but I'm open to it being rejected based on discussion.

### Context

Currently, there are two modules related to pcd loading in the `ndt_scan_matcher` package: `map_module` and `map_update_module`. These can be switched by setting `use_dynamic_map_loading` to true/false in the config. (Only one is used at a time).

* map_module : Has a callback for the map point cloud and sets the incoming point cloud to the NDT InputTarget
* map_update_module : Has a client for the pcd_loader service and dynamically loads the map. It also has ekf callback and timer callback.

Functionally, `map_update_module` is upwards compatible, so `map_module` is no longer necessary.

The `map_update_module` was introduced in pull request #2339 about a year ago. It has been working stably since then, so I believe now is a good time to fully migrate to using it instead of `map_module`.

### Changes

Removing `map_module` also requires `enable_differential_load` to always be true in map_loader, so this pull request also includes a fix for that.

If this pull request is acceptable, I will also modify autoware_launch to remove the parameters.

## Related PR
* https://github.com/autowarefoundation/autoware_launch/pull/769

## Tests performed

It has been confirmed that the `logging_simulator` runs with the same accuracy as before on AWSIM data with GT.

## Effects on system behavior

map_module will no longer be available. As the function is map_update_module, I don't think there will be any impact.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by mitsudome-r** [comment]
Does this mean that maps without map_config.yaml won't be supported after the removal of map_loader?

**Comment by SakodaShintaro** [comment]
@mitsudome-r 
Thank you for your comment.
I apologize, I was not aware of the existence of map_config.yaml. After some research, I understand that map_config.yaml is currently not implemented yet and may be used in the future.

I believe that as long as map_config.yaml is properly handled in the map loader, differential_map_loading will run normally. Depending on the implementation at the time, it will be easy to support even without map_config.yaml. 
I don't think the changes in this pull request will have any impact about map_config.yaml.

**Comment by Motsu-san** [comment]
> Does this mean that maps without map_config.yaml won't be supported after the removal of map_module?

@SakodaShintaro Simply you said, "No." didn't you?
@mitsudome-r Could you let us know where you worry? Is any slide needed to show that the changes in this pull request doesn't have any impact about map_config.yaml?

**Comment by SakodaShintaro** [comment]
@Motsu-san 
Yes. This pull request does not affect `map_config.yaml` in any way, I think.

**Comment by kminoda** [comment]
Assuming that @mitsudome-r's concern was about `pointcloud_map_metadata.yaml` and not `map_config.yaml`, the answer is no. Currently, the `differential_map_loading` I/F from `map_loader` will be successfully launched when either of the conditions is met:

1. The pcd is a single file
2. The pcd file consists of multiple files, and an appropriate `pointcloud_map_metadata.yaml` is provided

This does not change even if we disable the ROS 2 topic based I/F between `ndt_scan_matcher` and `map_loader`.

So, to summarize, a map without `pointcloud_map_metadata.yaml` will not be supported only if it consists of multiple PCD files, which is also true even before this PR.

**Comment by Motsu-san** [comment]
@mitsudome-r Does Minoda-san's explanation filled your concerns? Or should we provide an evidence(capture movie?) to prove the following?

> So, to summarize, a map without pointcloud_map_metadata.yaml will not be supported only if it consists of multiple PCD files, which is also true even before this PR.

**Comment by KYabuuchi** [comment]
related discussion: https://github.com/orgs/autowarefoundation/discussions/4062

File: localization/ndt_scan_matcher/README.md [hunk 1]
```diff
@@ -213,20 +213,15 @@ Using the feature, `ndt_scan_matcher` can theoretically handle any large size ma
...
+| `lidar_radius`                        | double | LiDAR radius used for localization (only used for diagnosis) |
 
 ### Enabling the dynamic map loading feature
 
-To use dynamic map loading feature for `ndt_scan_matcher`, you also need to appropriately configure some other settings outside of this node.
-Follow the next two instructions.
-
-1. enable dynamic map loading interface in `pointcloud_map_loader` (by setting `enable_differential_load` to true in the package)
-2. split the PCD files into grids (recommended size: 20[m] x 20[m])
+To use dynamic map loading feature for `ndt_scan_matcher`, you also need to split the PCD files into grids (recommended size: 20[m] x 20[m])
```
Comment (KYabuuchi) [inline localization/ndt_scan_matcher/README.md:224]: "dynamic_loading" is always enabled, so the title of this section seems strange to me.
How about "Notes for dynamic map loading"?

File: localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp [hunk 2]
```diff
@@ -410,10 +402,8 @@ void NDTScanMatcher::callback_initial_pose(
         << ". Please check the frame_id in the input topic and ensure it is correct.");
   }
 
-  if (use_dynamic_map_loading_) {
-    std::lock_guard<std::mutex> lock(latest_ekf_position_mtx_);
-    latest_ekf_position_ = initial_pose_msg_ptr->pose.pose.position;
-  }
+  std::lock_guard<std::mutex> lock(latest_ekf_position_mtx_);
+  latest_ekf_position_ = initial_pose_msg_ptr->pose.pose.position;
```
Comment (KYabuuchi) [inline localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:406]: How about adding a scope to emphasize that we are getting a lock?
I think this will make it less likely to be buggy when code is added behind it in the future.
```suggestion
  {
    std::lock_guard<std::mutex> lock(latest_ekf_position_mtx_);
    latest_ekf_position_ = initial_pose_msg_ptr->pose.pose.position;
  }
```

**Comment by KYabuuchi** [comment]
`pointcloud_map` and `input_ekf_odom` are no longer subscribed. So, please remove them in README.
image

image

Actually, input_ekf_odom was erased in [the other PR](https://github.com/autowarefoundation/autoware.universe/pull/5915)

File: localization/ndt_scan_matcher/config/ndt_scan_matcher.param.yaml
Comment (KYabuuchi) [inline localization/ndt_scan_matcher/config/ndt_scan_matcher.param.yaml:1]: it is necessary to create a PR in the [autoware_launch](https://github.com/autowarefoundation/autoware_launch) repository to remove `use_dynamic_map_loading` and `enable_differential_load` .

File: localization/ndt_scan_matcher/README.md (hunk 1 above)
Comment (SakodaShintaro) [reply localization/ndt_scan_matcher/README.md]: Fixed in https://github.com/autowarefoundation/autoware.universe/pull/5873/commits/a74c1ff2b1c5af5578ae94a4a1236c9c9570891c

File: localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp (hunk 2 above)
Comment (SakodaShintaro) [reply localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp]: Fixed in https://github.com/autowarefoundation/autoware.universe/pull/5873/commits/669ce13e6bbc6019d739c86bf5111d076cbaf5a0

File: localization/ndt_scan_matcher/config/ndt_scan_matcher.param.yaml
Comment (SakodaShintaro) [reply localization/ndt_scan_matcher/config/ndt_scan_matcher.param.yaml:1]: created
https://github.com/autowarefoundation/autoware_launch/pull/769

**Review by KYabuuchi — APPROVED** [review]
LGTM

**Comment by mitsudome-r** [comment]
@Motsu-san Thanks. I think @kminoda's explanation answers my concern.
